### PR TITLE
Optimizations for Paper (cargo network optimizations)

### PR DESCRIPTION
**Pull request by TheBusyBiscuit**
## Description

This Pull Request introduces a bunch of optimizations for Servers using [Paper](https://github.com/PaperMC/Paper).
Paper uses Asynchronous Chunk Generation/Loading which can improve performance for long-distance teleports.

### BlockState Snapshots
Normally, Spigot creates a new BlockState instance as a hard copy, for InventoryHolders the Inventory is only copied softly, as a reference. Since Inventories are already Thread-safe and only the reference is passed, cargo networks have no reason to create a BlockState snapshot. 
This behaviour cannot be avoided when using Spigot, with Paper it can.
So this will improve performance drastically (Sometimes up to 6 times smoother from what i tested) and optimize Cargo networks to work best when using Paper.

## Changes

* Added [PaperLib](https://github.com/PaperMC/PaperLib) as a shaded dependency
* Added the "Works best with Paper" suggestion on startup
* Optimized teleporter and elevators to use async-chunk-loading when possible
* Huge optimizations for Cargo networks and multiblocks to not take BlockState snapshots when accessing inventories

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/VersionsCommand.java [hunk 1]
```diff
@@ -33,7 +34,10 @@ public boolean isHidden() {
     @Override
     public void onExecute(CommandSender sender, String[] args) {
         if (sender.hasPermission("slimefun.command.versions") || sender instanceof ConsoleCommandSender) {
-            sender.sendMessage(ChatColors.color("&a" + Bukkit.getName() + " &2" + ReflectionUtils.getVersion()));
+            // After all these years... Spigot still displays as "CraftBukkit"
+            // so we will just this issue for them :)
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/VersionsCommand.java:38]: English?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 2]
```diff
@@ -134,8 +135,12 @@ public void open(Player p, Block b) {
                         yaw = -180 + (yaw - 180);
                     }
 
-                    player.teleport(new Location(player.getWorld(), block.getX() + 0.5, block.getY() + 0.4, block.getZ() + 0.5, yaw, player.getEyeLocation().getPitch()));
-                    player.sendTitle(ChatColor.WHITE + ChatColors.color(floor), " ", 20, 60, 20);
+                    Location destination = new Location(player.getWorld(), block.getX() + 0.5, block.getY() + 0.4, block.getZ() + 0.5, yaw, player.getEyeLocation().getPitch());
+                    PaperLib.teleportAsync(player, destination).thenAccept(teleported -> {
+                        if (teleported.booleanValue()) {
+                            Slimefun.runSync(() -> player.sendTitle(ChatColor.WHITE + ChatColors.color(floor), " ", 20, 60, 20));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:141]: These strings are nullable

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/VersionsCommand.java (hunk 1 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/VersionsCommand.java:38]: Ich entschuldige mich hierfür, es war sehr spät in der Nacht. ¯\_(ツ)_/¯

Yes, I can english good very.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java (hunk 2 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:141]: Which strings and why?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/gps/TeleportationManager.java [hunk 3]
```diff
@@ -58,10 +61,21 @@ public void openTeleporterGUI(Player p, UUID uuid, Block b, int complexity) {
...
+                List<String> nameAndLore = new ArrayList<>();
+                nameAndLore.add(waypoint.getName().replace("player:death ", ""));
+                nameAndLore.add("");
+                nameAndLore.add("&8\u21E8 &7" + SlimefunPlugin.getLocalization().getResourceString(p, "tooltips.world") + ": &f" + l.getWorld().getName());
+                nameAndLore.add("&8\u21E8 &7X: &f" + l.getX());
+                nameAndLore.add("&8\u21E8 &7Y: &f" + l.getY());
+                nameAndLore.add("&8\u21E8 &7Z: &f" + l.getZ());
+                nameAndLore.add("&8\u21E8 &7" + SlimefunPlugin.getLocalization().getMessage(p, "machines.TELEPORTER.gui.time") + ": &f" + time + "s");
+                nameAndLore.add("");
+                nameAndLore.add("&8\u21E8 &c" + SlimefunPlugin.getLocalization().getMessage(p, "machines.TELEPORTER.gui.tooltip"));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/gps/TeleportationManager.java:76]: Probably best to use Arrays.asList

I can't see a reason why this should be done. Would be slower and means there's a mutable list (even if for a short time)

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java [hunk 4]
```diff
@@ -134,8 +135,12 @@ public void open(Player p, Block b) {
                         yaw = -180 + (yaw - 180);
                     }
 
-                    player.teleport(new Location(player.getWorld(), block.getX() + 0.5, block.getY() + 0.4, block.getZ() + 0.5, yaw, player.getEyeLocation().getPitch()));
-                    player.sendTitle(ChatColor.WHITE + ChatColors.color(floor), " ", 20, 60, 20);
+                    Location destination = new Location(player.getWorld(), block.getX() + 0.5, block.getY() + 0.4, block.getZ() + 0.5, yaw, player.getEyeLocation().getPitch());
+                    PaperLib.teleportAsync(player, destination).thenAccept(teleported -> {
+                        if (teleported.booleanValue()) {
+                            Slimefun.runSync(() -> player.sendTitle(ChatColor.WHITE + ChatColors.color(floor), null, 20, 60, 20));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/gps/ElevatorPlate.java:141]: Check if this can be async first.

**Comment by TheBusyBiscuit** [comment]
Ready to be merged? @WalshyDev 

I tested it on Spigot and Paper again, no issues thus far.

**Pull request merged**